### ISSUE TITLE
#[forbid(unreachable_code)] incompatible with the Try ('?') operator

### ISSUE THREAD

**Issue by Dark-Legion** [rust-lang/rust#76053]
I tried this code:

```rust
#![forbid(unreachable_code)]

#[allow(dead_code)] // Remove warning.
fn foo() -> Option<()> {
    Some(0)?;
    None // Just so it doesn't emit error here.
}
```

Both on debug and release it generates an error saying that `#[allow(unreachable_code)]` was overridden by the `#[forbid(...)]` attribute, while pointing at the "try" expression.

### Meta

`rustc --version --verbose`:
```
rustc 1.46.0 (04488afe3 2020-08-24)
binary: rustc
commit-hash: 04488afe34512aa4c33566eb16d8c912a3ae04f9
commit-date: 2020-08-24
host: x86_64-pc-windows-msvc
release: 1.46.0
LLVM version: 10.0
```

**Comment by LeSeulArtichaut** [comment]
According to Godbolt this is the case since 1.22.0 where the `Try` trait was implemented on `Option`.

**Comment by LeSeulArtichaut** [comment]
This bug can also be reproduced with a `Result`:

```rust
#![forbid(unreachable_code)]

#[allow(dead_code)] // Remove warning.
fn foo() -> Result<(), ()> {
    Ok(0)?;
    Err(()) // Just so it doesn't emit error here.
}
```

Which allows to find out that this regressed in 1.16.0.

**Comment by spastorino** [comment]
Assigning `P-low` as [discussed as part of the Prioritization Working Group procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/I-prioritize.20.2376053.20.23.5Bforbid(unreachable_code).5D.20incompatibl.E2.80.A6) and removing `I-prioritize`.

**Comment by Dylan-DPC** [comment]
This code now compiles fine, we just need a test added against this to close the issue

**Comment by RodBurman** [comment]
Agreed as or release 1.83 the code compiles with no errors or warnings. Has the test been added?
